Task05/src/main.cpp

```diff
@@ -0,0 +1,227 @@
+}
+
+/**
+ * Calculates the Chebyshev distance between two points (x1, y1) and (x2, y2).
+ */
+double calculateDistanceLinf(double x1, double y1, double x2, double y2) {
+    return std::max(std::abs(x1 - x2), std::abs(y1 - y2));
+}
+
+/**
+ * Calculates the Manhattan distance between two points (x1, y1) and (x2, y2).
+ */
+double calculateDistanceL1(double x1, double y1, double x2, double y2) {
+    return std::abs(x1 - x2) + std::abs(y1 - y2);
+}
+
+/**
+ * Reads the city data from a file and returns an unordered map containing the city names and their coordinates.
+ * The file format should be:
+ *   - Line 1: City name
+ *   - Line 2: City coordinates (x-y format)
+ *   - Repeat for each city
+ */
+std::unordered_map<std::string, std::pair<double, double>> readCitiesFromFile(const std::string& filename) {
+    std::ifstream file(filename);
+    std::unordered_map<std::string, std::pair<double, double>> cities;
+
+    if (!file.is_open()) {
+        throw std::runtime_error("Error opening file: " + filename);
+    }
+
+    // Check if the file is empty
+    file.seekg(0, std::ios::end);
+    if (file.tellg() == 0) {
+        throw std::runtime_error("Empty file: " + filename);
+    }
+    file.seekg(0, std::ios::beg);
+
+    std::string line;
+    std::string cityName;
+    std::string cityCoordinates;
+
+    while (std::getline(file, line)) {
+        cityName = line;
+        if (!std::getline(file, line)) {
+            throw std::runtime_error("Invalid file format: " + filename);
+        }
+        cityCoordinates = line;
+
+        size_t pos = cityCoordinates.find('-');
+        if (pos == std::string::npos) {
+            throw std::runtime_error("Invalid file format: " + filename);
+        }
+
+        double x, y;
+        char hyphen;
+        std::istringstream iss(cityCoordinates);
+        if (!(iss >> x >> hyphen >> y)) {
+            throw std::runtime_error("Invalid input for coordinates in file: " + filename);
+        }
+
+        cities[cityName] = std::make_pair(x, y);
+    }
+
+    file.close();
+    return cities;
+}
+
+/**
+ * Searches for cities within a given radius around a selected city.
+ * Prints the nearby cities and the number of cities to the north of the selected city.
+ * The distance function used for the search is determined by the norm parameter.
+ */
+void searchCities(const std::unordered_map<std::string, std::pair<double, double>>& cities, const std::string& cityName, double radius, const std::function<double(double, double, double, double)>& distanceFunc) {
+    std::unordered_map<std::string, std::pair<double, double>> nearbyCities;
+    std::unordered_map<std::string, std::pair<double, double>> northCities;
+
+    const auto& selectedCity = cities.find(cityName);
+    if (selectedCity == cities.end()) {
+        std::cout << "ERROR: \"" << cityName << "\" isn't found in the city list. Please try again." << std::endl;
+        return;
+    }
+    nearbyCities.reserve(cities.size());
```

Comment:
you are missing 1 step in finding the cities in radius, you should find vertical and horizontal rectangles, and then to find the intersection square, and to find a circle inside the square.